openscenario/map_fragment/include/map_fragment/road_segment.hpp

```diff
@@ -0,0 +1,155 @@
+    {
+      throw std::invalid_argument(
+        "Expected lane_width to be positive. Actual value: "
+        + std::to_string(lane_width)
+      );
+    }
+  }
+}; // struct RoadCrossSectionDescription
+
+class RoadCrossSection
+{
+  lanelet::Points3d points_;
+
+public:
+  explicit RoadCrossSection(RoadCrossSectionDescription const& description,
+                            Point2d const& origin,
+                            Vector2d const& tangent_vector)
+  {
+    auto n = description.number_of_lanes;
+    auto w = description.lane_width;
+    auto normal_vector = rotate(tangent_vector, M_PI / 2);
+
+    for (auto i = 0; i < description.number_of_lanes + 1; i++)
+    {
+      auto lateral_position = (i - n / 2) * w;
+      auto p = origin + normal_vector * lateral_position;
+      points_.push_back(makePoint3d(p.x, p.y, 0.));
+    }
+  }
+
+  auto getPoints() -> const lanelet::Points3d&
+  {
+    return points_;
+  }
+}; // class RoadCrossSection
+
+class RoadSegment
+{
+  ParametricCurve::Pointer guide_curve_;
+  RoadCrossSectionDescription cross_section_description_;
+
+public:
+  explicit RoadSegment(ParametricCurve::Pointer guide_curve,
+                       RoadCrossSectionDescription const& cross_section_description)
+    : guide_curve_(guide_curve)
+    , cross_section_description_(cross_section_description)
+  {
+  }
+
+  auto getLanelets(double resolution) -> const lanelet::Lanelets
+  {
+    lanelet::Lanelets lanelets;
+    auto n = cross_section_description_.number_of_lanes + 1;
+
+    // TODO Find a cleaner way to do this
+    std::vector<lanelet::LineString3d> lane_boundaries(n);
+    for (auto i = 0; i < n; i++)
+    {
+      lane_boundaries[i].setId(i);
+    }
+
+    for (auto i = 0; i < resolution; i++)
+    {
+      auto t = i / (resolution - 1);
+      auto position = guide_curve_->getPosition(t);
+      auto tangent_vector = guide_curve_->getTangentVector(t);
+
+      RoadCrossSection cross_section(cross_section_description_, position, tangent_vector);
+      auto cross_section_points = cross_section.getPoints();
+
+      for (auto j = 0; j < cross_section_description_.number_of_lanes + 1; j++)
+      {
+        lane_boundaries[j].push_back(cross_section_points[j]);
+      }
+    }
+
+    for (auto i = 0; i < cross_section_description_.number_of_lanes; i++)
+    {
+      lanelets.push_back(makeLanelet(
+        lane_boundaries[i],
+        lane_boundaries[i + 1]));
+    }
+
+    return lanelets;
+  }
+}; // class RoadSegment
+
+auto makeCurvedRoadSegment(double curvature,
+                           double length,
+                           int number_of_lanes,
+                           double lane_width) -> RoadSegment
+{
+  ParametricCurve::Pointer guide_curve;
+
+  if (curvature == 0.0)
+  {
+    guide_curve = std::make_shared<Straight>(length);
+  }
+  else
+  {
+    auto radius = std::abs(1 / curvature);
+    auto angle = length * curvature;
+    guide_curve = std::make_shared<Arc>(radius, angle);
+  }
```

Comment:
```suggestion
  const ParametricCurve::Pointer guide_curve =
    curvature == 0.0 ? std::make_shared<Straight>(length)
                     : std::make_shared<Arc>(std::abs(1 / curvature), length * curvature);
```

(This suggested code will probably be complained of code style divergence from `ament_clang_format`.)

Reply:
The suggested code results in compiler errors:

```
--- stderr: map_fragment                                   
In file included from /home/robotec/Code/RJD-617/workspace/src/scenario_simulator_v2/openscenario/map_fragment/src/lane.cpp:16:
/home/robotec/Code/RJD-617/workspace/src/scenario_simulator_v2/openscenario/map_fragment/include/map_fragment/road_segment.hpp: In function ‘map_fragment::RoadSegment map_fragment::makeCurvedRoadSegment(double, double, int, double)’:
/home/robotec/Code/RJD-617/workspace/src/scenario_simulator_v2/openscenario/map_fragment/include/map_fragment/road_segment.hpp:120:22: error: operands to ‘?:’ have different types ‘std::shared_ptr<map_fragment::Straight>’ and ‘std::shared_ptr<map_fragment::Arc>’
  120 |     curvature == 0.0 ? std::make_shared<Straight>(length)
      |     ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  121 |                      : std::make_shared<Arc>(std::abs(1 / curvature), length * curvature);
      |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/robotec/Code/RJD-617/workspace/src/scenario_simulator_v2/openscenario/map_fragment/src/curve.cpp:16:
/home/robotec/Code/RJD-617/workspace/src/scenario_simulator_v2/openscenario/map_fragment/include/map_fragment/road_segment.hpp: In function ‘map_fragment::RoadSegment map_fragment::makeCurvedRoadSegment(double, double, int, double)’:
/home/robotec/Code/RJD-617/workspace/src/scenario_simulator_v2/openscenario/map_fragment/include/map_fragment/road_segment.hpp:120:22: error: operands to ‘?:’ have different types ‘std::shared_ptr<map_fragment::Straight>’ and ‘std::shared_ptr<map_fragment::Arc>’
  120 |     curvature == 0.0 ? std::make_shared<Straight>(length)
      |     ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  121 |                      : std::make_shared<Arc>(std::abs(1 / curvature), length * curvature);
      |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [CMakeFiles/curve.dir/build.make:76: CMakeFiles/curve.dir/src/curve.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:149: CMakeFiles/curve.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
gmake[2]: *** [CMakeFiles/lane.dir/build.make:76: CMakeFiles/lane.dir/src/lane.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:227: CMakeFiles/lane.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
```
This can be solved with `static_cast`s:

```C++
const auto guide_curve =
  curvature == 0.0 ? static_cast<ParametricCurve::Pointer>(std::make_shared<Straight>(length))
                   : static_cast<ParametricCurve::Pointer>(
                       std::make_shared<Arc>(std::abs(1 / curvature), length * curvature));
```

or:

```C++
const auto guide_curve =
  curvature == 0.0 ? static_cast<ParametricCurve::Pointer>( \
                       std::make_shared<Straight>(length))
                   : static_cast<ParametricCurve::Pointer>(
                       std::make_shared<Arc>(std::abs(1 / curvature), length * curvature));
```

I think I prefer the second one.

Reply:
(I'm sorry for suggesting code that I haven't confirmed will compile.)

Problems may occur if you do not use std::static_pointer_cast instead of static_cast.

Reply:
[Fixed](https://github.com/tier4/scenario_simulator_v2/pull/1108/commits/d54368e141acefa9f82b3f64c2e8890950073483).